tests/index_routines.py

```diff
@@ -92,6 +99,43 @@ def test():
                     num_res = num.choose(num_choices, (num_rhs1, num_rhs2))
                     assert np.array_equal(np_res, num_res)
 
+    # --------------------------------------------------------------
+    # diagonal operator
+    # --------------------------------------------------------------
+    ad = np.arange(24).reshape(4, 3, 2)
+    num_ad = num.array(ad)
+    assert np.array_equal(ad.diagonal(), num_ad.diagonal())
+    assert np.array_equal(ad.diagonal(0, 1, 2), num_ad.diagonal(0, 1, 2))
+    assert np.array_equal(ad.diagonal(1, 0, 2), num_ad.diagonal(1, 0, 2))
+
```

Comment:
Can you add cases for:
* extract=False (offset=0, offset>0, offset<0)
* extract=True (offset>0, offset<0)

Reply:
Do you want me to explicitly specify extract when calling `diagonal`?
I believe the first use case is tested here:
```
for offset in range(-diag_size + 1, diag_size):
                assert np.array_equal(
                    np_array.diagonal(offset, axes[0], axes[1]),
                    num_array.diagonal(offset, axes[0], axes[1]),
                )
```

And the second use case is tested through `diag` in diag.py

BTW, do you think I should just move tests from diag.py to index_toutines.py? Or vice versa ?

Reply:
Sounds like all cases are covered already, I don't think we need to add more tests. Let's just move the tests from diag.py to index_routines.py.